core/sql/sqlcomp/CmpSeabaseDDLroutine.cpp

```diff
@@ -1925,3 +2730,568 @@ short CmpSeabaseDDL::createSeabaseLibmgrCPPLib(ExeCliInterface * cliInterface)
+        
+        mdui->setMsg("Upgrade Libraries: Drop Old Libraries done");
+        mdui->setEndStep(TRUE);
+        mdui->setSubstep(0);
+         
+        return 0;
+      }
+      break;
+
+    default:
+      return -1;
+    }
+
+return 0;
+}
+
+
+short CmpSeabaseDDL::upgradeLibrariesUndo(ExeCliInterface * cliInterface,
+                                  CmpDDLwithStatusInfo *mdui)
+{
+  Lng32 cliRC = 0;
+
+  while (1) // exit via return stmt in switch
+    {
+      switch (mdui->subStep())
+        {
+        // error return codes from upgradeLibraries can be mapped to
+        // the right recovery substep by this formula: substep = -(retcode + 1)
+        case 0: // corresponds to -1 return code from upgradeRepos (or
+                // to full recovery after some error after upgradeRepos)
+        case 1: // corresponds to -2 return code from upgradeRepos
+        case 2: // corresponds to -3 return code from upgradeRepos
+          {
+            mdui->setMsg("Upgrade Libraries: Restoring Old Libraries");
+            mdui->setSubstep(2*mdui->subStep()+3); // go to appropriate case
+            mdui->setEndStep(FALSE);
+        
+            return 0;
+          }
+          break;
+
+        case 3:
+          {
+            mdui->setMsg(" Start: Drop New Libraries");
+            mdui->subStep()++;
+            mdui->setEndStep(FALSE);
+        
+            return 0;
+          }
+          break;
+
+        case 4:
+          {
+            // drop new Libraries; ignore errors
+            dropLibraries(cliInterface, FALSE/*new repos*/, 
+                          TRUE /* don't drop new tables that haven't been upgraded */);
+            cliInterface->clearGlobalDiags();
+            mdui->setMsg(" End: Drop New Libraries");
+            mdui->subStep()++;
+            mdui->setEndStep(FALSE);
+        
+            return 0;
+          }
+          break;
+
+        case 5:
+          {
+            mdui->setMsg(" Start: Rename Old Libraries back to New");
+            mdui->subStep()++;
+            mdui->setEndStep(FALSE);
+        
+            return 0;
+          }
+          break;
+ 
+        case 6:
+          {
+            // rename old Libraries to current; ignore errors
+            alterRenameLibraries(cliInterface, FALSE);
+            cliInterface->clearGlobalDiags();
+            mdui->setMsg(" End: Rename Old Libraries back to New");
+            mdui->subStep()++;
+            mdui->setEndStep(FALSE);
+        
+            return 0;
+          }
+          break;
+
+        case 7:
+          {
+            mdui->setMsg("Upgrade Libraries: Restore done");
+            mdui->setSubstep(0);
+            mdui->setEndStep(TRUE);
+        
+            return 0;
+          }
+          break;
+
+        default:
+          return -1;
+        }
+    } // while
+
+  return 0;
+
+}
+
+short CmpSeabaseDDL::createLibraries(ExeCliInterface * cliInterface)
```

Comment:
This name confused me somewhat. As far as I understand it now, this creates the new ```"_MD_".LIBRARIES``` table, but not any libraries.

Reply:
Yes this is creating the LIBRARIES table only.